File: server/internal/repositories/redis_jobs_repository.go
package repositories

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
)

type RedisJobsRepository interface {
	CreateJobMeta(ctx context.Context, jobID string, total int, ttlSeconds int) error
	PushQueue(ctx context.Context, queue string, payload any) error
	GetMeta(ctx context.Context, jobID string) (map[string]string, error)
	GetResults(ctx context.Context, jobID string) ([]string, error)
	GetErrors(ctx context.Context, jobID string) ([]string, error)
}

type redisJobsRepositoryImpl struct{ rdb *redis.Client }

func NewRedisJobsRepository(rdb *redis.Client) RedisJobsRepository {
	return &redisJobsRepositoryImpl{rdb: rdb}
}

func metaKey(jobID string) string { return fmt.Sprintf("job:%s:meta", jobID) }

func resultsKey(jobID string) string { return fmt.Sprintf("job:%s:results", jobID) }

func errorsKey(jobID string) string { return fmt.Sprintf("job:%s:errors", jobID) }

func (r *redisJobsRepositoryImpl) CreateJobMeta(ctx context.Context, jobID string, total int, ttlSeconds int) error {
	mk := metaKey(jobID)

	pipe := r.rdb.Pipeline()

	pipe.HSet(ctx, mk, map[string]any{"status": "QUEUED", "total": total, "processed": 0, "failed": 0})

	ttl := time.Duration(ttlSeconds) * time.Second
	pipe.Expire(ctx, mk, ttl)
	pipe.Expire(ctx, resultsKey(jobID), ttl)
	pipe.Expire(ctx, errorsKey(jobID), ttl)

	_, err := pipe.Exec(ctx)

	return err
}

func (r *redisJobsRepositoryImpl) PushQueue(ctx context.Context, queue string, payload any) error {
	b, err := json.Marshal(payload)
	if err != nil {
		return err
	} // LPUSH para que el worker BRPOP lo tome
	return r.rdb.LPush(ctx, queue, string(b)).Err()
}

func (r *redisJobsRepositoryImpl) GetMeta(ctx context.Context, jobID string) (map[string]string, error) {
	return r.rdb.HGetAll(ctx, metaKey(jobID)).Result()
}

func (r *redisJobsRepositoryImpl) GetResults(ctx context.Context, jobID string) ([]string, error) {
	return r.rdb.LRange(ctx, resultsKey(jobID), 0, -1).Result()
}

func (r *redisJobsRepositoryImpl) GetErrors(ctx context.Context, jobID string) ([]string, error) {
	return r.rdb.LRange(ctx, errorsKey(jobID), 0, -1).Result()
}

File: pdf-service-test/app/repositories/redis_jobs_repository.py
from __future__ import annotations
import json
from typing import Any
from redis.asyncio import Redis

from app.core.config import Settings
from app.repositories.jobs_repository import JobsRepository

class RedisJobsRepository(JobsRepository):
    def __init__(self, redis: Redis, settings: Settings) -> None:
        self.redis = redis
        self.settings = settings

    def _meta_key(self, job_id: str) -> str:
        return f"job:{job_id}:meta"

    def _results_key(self, job_id: str) -> str:
        return f"job:{job_id}:results"

    def _errors_key(self, job_id: str) -> str:
        return f"job:{job_id}:errors"

    async def create_job(self, job_id: str, total: int) -> None:
        meta_key = self._meta_key(job_id)
        await self.redis.hset(meta_key, mapping={
            "status": "QUEUED",
            "total": total,
            "processed": 0,
            "failed": 0,
        })
        ttl = int(self.settings.REDIS_JOB_TTL_SECONDS)
        await self.redis.expire(meta_key, ttl)
        await self.redis.expire(self._results_key(job_id), ttl)
        await self.redis.expire(self._errors_key(job_id), ttl)

    async def push_job(self, payload: dict[str, Any]) -> None:
        queue_key = self.settings.REDIS_QUEUE_PDF_JOBS
        await self.redis.lpush(queue_key, json.dumps(payload))

    async def get_meta(self, job_id: str) -> dict[str, Any]:
        return await self.redis.hgetall(self._meta_key(job_id))

    async def get_results(self, job_id: str, start: int = 0, end: int = -1) -> list[str]:
        return await self.redis.lrange(self._results_key(job_id), start, end)
